src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs

```diff
@@ -100,38 +100,41 @@ public IEnumerable<TState> GetStatesInCompilationOrder()
 
         public ImmutableArray<TValue> SelectAsArray<TValue>(Func<TState, TValue> selector)
         {
-            using var _ = ArrayBuilder<TValue>.GetInstance(out var builder);
+            // Directly use ImmutableArray.Builder as we know the final size
+            var builder = ImmutableArray.CreateBuilder<TValue>(_map.Count);
```

Comment:
Previously this was reusing instances from a pool. Now, this will allocate each time.

Reply:
A couple thoughts:

1) Previously, the code called builder.ToImmutable every time, which allocates. 
2) Previously the code would incrementally resize by calling Add. This could cause multiple re-allocations before reaching it's final size. (and have it's final size be up to twice too large)
3) Previously the code would end up not adding the builder back to the pool if it exceeded the size threshold (fairly common in the scenario I'm testing)
4) Previously, the code would need to do the overhead of searching the ObjectPool for an object to use.

Reply:
Makes sense!